### ISSUE TITLE
Collective moderation

### ISSUE THREAD

**Issue by Betree** [CaptainFact/captain-fact-frontend#24]
# Description

Comments with a certain number of flags get reported and are no more returned from the API. Modifications, such as editing a statement or adding a speaker can also be flagged.

To ensure a collective moderation we should create a moderation platform reserved to users with a huge reputation.

## Moderators privileges

For each reported comments, moderators have 3 choices:

* Confirm flag => +1
* Not sure => 0
* Abusive flag => -1

## Rules

We need at least 3 moderators to vote before taking a decision. Then we generate a score:

```
score = (nb_confirm - nb_abuse) / total_feedbacks
```
<br/>
Which generates a score between -1.0 and +1.0.

* If `score <= -0.66` => Abusive flag 
* If `score >=  0.66` => Confirm flag 
* Else => Wait to have more feedbacks 

<details>
<summary>Examples</summary>
  <ul>
    <li>2 confirm + 1 not sure = 2 * 1 + 0 = 2 => Confirm</li>
    <li>1 confirm + 2 not sure = 1 + 0 + 0 = 1 => Re-ask with more moderators</li>
    <li>1 confirm + 2 abusive = 1 - 1 - 1 = -1 => Re-ask with more moderators</li>
    <li>2 abusive + 1 not sure = 2 * (-1) + 0 = -2 => Abusive flag</li>
  </ul>
</details>

## Reputation updates

We update reputations depending on this score. But to be fair and not penalize users when decisions are complicated,
we first determine the strength of the moderators consensus with: 

```
consensur_strength = (abs_score - min_val) * (1 / (1 - min_val))
```
<br/>
Which gives us a value between 0.0 and 1.0 that we use to ponderate reputatin gain or loss.

</details>

## Moderators reward

* A special achievement
* Moderators should gain additional reputation (not too much) when they take the time to check flags

# Interface

Moderation platform should take two shapes:

1. Check all flagged comments and actions linked to a video
2. A page where we would get a random action or comment assigned, one at a time

For each action, moderator should be presented a box like this one (extended version [here](https://github.com/CaptainFact/captain-fact-prototyping/blob/collective-moderation/CaptainFact.io/Moderation/Action%20(extended).png?raw=true)):
![Action moderation preview](https://github.com/CaptainFact/captain-fact-prototyping/blob/collective-moderation/CaptainFact.io/Moderation/Action.png?raw=true)

This is pretty much what we show on the history page without the Approve / Flag buttons on actions.

* Full page prototype with random actions

![Full page prototype with random actions](https://github.com/CaptainFact/captain-fact-prototyping/blob/collective-moderation/CaptainFact.io/Moderation/Random.png?raw=true)

* (Bonus) Full page prototype on video collective moderation

![Full page prototype on video collective moderation](https://github.com/CaptainFact/captain-fact-prototyping/blob/collective-moderation/CaptainFact.io/Moderation/Video.png?raw=true)

# Code

## Frontend

* `app/constants.js` => Some action types and entities are defined here.
  If you want to add Comment entity: `export const ENTITY_COMMENT = 4`
* `assets/assets/locales/en/history.json` => Actions / Entities translations indexed by their numeric identifier. Again **comment type is 4** - you can add it there.

## API endpoints (REST)

Methods presented in details deal with actions. We're always posting feedback on an action (ex: "create comment") and not on the entity directly (comment). An action looks like that:

```json
{
    "user": {
      "id": 2126,
      ...
    },
    "id": 1972,
    "type": 1, // Action type (create, remove, update...etc)
    "entity": 4, // Entity (comment, statement, video...etc)
    "entity_id": 538,
    "time": "2017-11-12T04:26:12.573219",
    "changes": {
      // Changes made with this action
      "text": "I like to move it"
    }
  }
```

<details>

##### Request
* __Method:__ GET
* __Path:__ /moderation/videos/:video_id

##### Response
* __Status__: 200
* __Response body:__
```json
[
  {
    "user": {
      "username": "User-G4R8umEp45",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.569109",
      "picture_url": null,
      "name": "Nella",
      "mini_picture_url": null,
      "id": 2126,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.573219",
    "id": 1972,
    "entity_id": 538,
    "entity": 4,
    "changes": {
      "text": "I like to move it"
    }
  },
  {
    "user": {
      "username": "User-546rUsidyO",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.574333",
      "picture_url": null,
      "name": "Junior",
      "mini_picture_url": null,
      "id": 2128,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.575786",
    "id": 1973,
    "entity_id": 539,
    "entity": 4,
    "changes": {
      "text": "I like to move it"
    }
  },
  {
    "user": {
      "username": "User-VwiSbNEwv-",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.576610",
      "picture_url": null,
      "name": "Patricia",
      "mini_picture_url": null,
      "id": 2130,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.578125",
    "id": 1974,
    "entity_id": 540,
    "entity": 4,
    "changes": {
      "text": "I like to move it"
    }
  },
  {
    "user": {
      "username": "User-GcfGMKNkx7",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.579005",
      "picture_url": null,
      "name": "Abigail",
      "mini_picture_url": null,
      "id": 2132,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.580488",
    "id": 1975,
    "entity_id": 541,
    "entity": 4,
    "changes": {
      "text": "I like to move it"
    }
  },
  {
    "user": {
      "username": "User-JxQBL38e_W",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.581350",
      "picture_url": null,
      "name": "Mason",
      "mini_picture_url": null,
      "id": 2134,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.583214",
    "id": 1976,
    "entity_id": 542,
    "entity": 4,
    "changes": {
      "text": "I like to move it"
    }
  }
]
```

##### Request
* __Method:__ GET
* __Path:__ /moderation/random
* __Params:__ (optional) `count` => number of actions to return (default: 5)
##### Response
* __Status__: 200
* __Response body:__
```json
[
  {
    "user": {
      "username": "User-IVrwM7LAhj",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.519572",
      "picture_url": null,
      "name": "Florencio",
      "mini_picture_url": null,
      "id": 2114,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.521282",
    "id": 1961,
    "entity_id": 535,
    "entity": 4,
    "changes": {
      "text": "Jouje is the best jouje"
    }
  },
  {
    "user": {
      "username": "User-fyflKOAHFL",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.515274",
      "picture_url": null,
      "name": "Rosalyn",
      "mini_picture_url": null,
      "id": 2112,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.517785",
    "id": 1960,
    "entity_id": 534,
    "entity": 4,
    "changes": {
      "text": "I like to move it...."
    }
  },
  {
    "user": {
      "username": "User-MnHUvqvy01",
      "reputation": 0,
      "registered_at": "2017-11-12T04:26:12.523268",
      "picture_url": null,
      "name": "Ora",
      "mini_picture_url": null,
      "id": 2116,
      "achievements": [
        1
      ]
    },
    "type": 1,
    "time": "2017-11-12T04:26:12.524697",
    "id": 1962,
    "entity_id": 536,
    "entity": 4,
    "changes": {
      "text": "No problemo, he said"
    }
  }
]
```

##### Request
* __Method:__ POST
* __Path:__ /moderation/feedback
* __Request body:__
```json
{
  "value": 1,
  "action_id": 1969
}
```
##### Response
* __Status__: 204

</details>

# Resources

Stackoverflow blog posts about collective moderation:

* https://stackoverflow.blog/2009/05/18/a-theory-of-moderation/
* https://stackoverflow.blog/2010/07/27/moderator-pro-tempore/
* https://stackoverflow.blog/2010/12/02/stack-exchange-moderator-elections-begin/

**Comment by Betree** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Community moderation](https://trello.com/c/s1EOyDQM/425-community-moderation)

**Comment by Betree** [comment]
API is ready. Documentation above has been updated, see `API endpoints (REST) > Details`

**Comment by Betree** [comment]
Implemented by #79 🍾🎇

Thanks @NGambini !
